# **Titanic Dataset**
*It is one of the most popular datasets used for understanding machine learning basics. It contains information of all the passengers aboard the RMS Titanic, which unfortunately was shipwrecked. This dataset can be used to predict whether a given passenger survived or not.*

#**Importing the Dataset**

In [ ]:
#Importing Pandas Library
import pandas as pd

#Loading Data
titanic = pd.read_csv('https://raw.githubusercontent.com/eclarson/DataMiningNotebooks/master/data/titanic.csv')

#**Basic Operations**

##*Print the first few records of the file to look at the data*

###**Method 1 :** By Default
###*(In colab By Default only First 12 Rows are Printed, we Do need to know the Dimension of dataset beforehand)*

In [ ]:
titanic

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C


###**Method 2 :** By Slicing.
###*(It gives us More Flexibilty, To Select Number of Rows and the Index we wish to See, Only Downside is that we need to know the dimensions Beforehand)*

In [ ]:
titanic[:10]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
7,8,0,3,"Palsson, Master. Gosta Leonard",male,2.0,3,1,349909,21.0750,NaN,S
8,9,1,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female,27.0,0,2,347742,11.1333,NaN,S
9,10,1,2,"Nasser, Mrs. Nicholas (Adele Achem)",female,14.0,1,0,237736,30.0708,NaN,C


###**Method 3 :** Pandas head() Method.
###*(Pandas head() method is used to return top n (5 by default) rows of a data frame or series).*

In [ ]:
titanic.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [ ]:
titanic.head(10)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
7,8,0,3,"Palsson, Master. Gosta Leonard",male,2.0,3,1,349909,21.0750,NaN,S
8,9,1,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female,27.0,0,2,347742,11.1333,NaN,S
9,10,1,2,"Nasser, Mrs. Nicholas (Adele Achem)",female,14.0,1,0,237736,30.0708,NaN,C


###**Method 4 :** Using Pandas iloc[] Method.
###*(The Pandas library provides a unique method to retrieve rows from a Data Frame. Dataframe.iloc[] method is used when the index label of a data frame is something other than numeric series of 0, 1, 2, 3….n or in case the user doesn’t know the index label. Rows can be extracted using an imaginary index position which isn’t visible in the data frame.)*

In [ ]:
titanic.iloc[:10]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
7,8,0,3,"Palsson, Master. Gosta Leonard",male,2.0,3,1,349909,21.0750,NaN,S
8,9,1,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female,27.0,0,2,347742,11.1333,NaN,S
9,10,1,2,"Nasser, Mrs. Nicholas (Adele Achem)",female,14.0,1,0,237736,30.0708,NaN,C


###**Method 5 :** Display Records of Specific Colunmn

In [ ]:
titanic[['PassengerId', 'Survived','Name']].head(10)

,PassengerId,Survived,Name
0,1,0,"Braund, Mr. Owen Harris"
1,2,1,"Cumings, Mrs. John Bradley (Florence Briggs Th..."
2,3,1,"Heikkinen, Miss. Laina"
3,4,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)"
4,5,0,"Allen, Mr. William Henry"
5,6,0,"Moran, Mr. James"
6,7,0,"McCarthy, Mr. Timothy J"
7,8,0,"Palsson, Master. Gosta Leonard"
8,9,1,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)"
9,10,1,"Nasser, Mrs. Nicholas (Adele Achem)"


##Print some Basic Statistics

###**Method 1 :** using Pandas describe() method.
###*(Pandas describe() is used to view some basic statistical details like percentile, mean, std etc. of a data frame or a series of numeric values. When this method is applied to a series of string, it returns a different output which is shown in the examples below).*

In [ ]:
titanic.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


###**Method 2 :** Pandas mean() Method.
###*Return the mean of the values over the requested axis.*

In [ ]:
titanic['Age'].mean()

29.69911764705882

In [ ]:
titanic.mean(numeric_only = True)

,0
PassengerId,446.000000
Survived,0.383838
Pclass,2.308642
Age,29.699118
SibSp,0.523008
Parch,0.381594
Fare,32.204208


###**Method 3 :** Pandas median() Method

In [ ]:
titanic.median(numeric_only = True)

,0
PassengerId,446.0000
Survived,0.0000
Pclass,3.0000
Age,28.0000
SibSp,0.0000
Parch,0.0000
Fare,14.4542


###**Method 4 :** Pandas mode() Method


In [ ]:
titanic.mode(axis=0)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0.0,3.0,"Abbing, Mr. Anthony",male,24.0,0.0,0.0,1601,8.05,B96 B98,S
1,2,NaN,NaN,"Abbott, Mr. Rossmore Edward",NaN,NaN,NaN,NaN,347082,NaN,C23 C25 C27,NaN
2,3,NaN,NaN,"Abbott, Mrs. Stanton (Rosa Hunt)",NaN,NaN,NaN,NaN,CA. 2343,NaN,G6,NaN
3,4,NaN,NaN,"Abelson, Mr. Samuel",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,5,NaN,NaN,"Abelson, Mrs. Samuel (Hannah Wizosky)",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,NaN,NaN,"de Mulder, Mr. Theodore",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
887,888,NaN,NaN,"de Pelsmaeker, Mr. Alfons",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
888,889,NaN,NaN,"del Carlo, Mr. Sebastiano",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
889,890,NaN,NaN,"van Billiard, Mr. Austin Blyler",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


###**Method 5 :** Pandas count() Method.

In [ ]:
titanic.count()

,0
PassengerId,891
Survived,891
Pclass,891
Name,891
Sex,891
Age,714
SibSp,891
Parch,891
Ticket,891
Fare,891


###**Method 6 :** Pandas max() Method.


In [ ]:
titanic.max(numeric_only = True)

,0
PassengerId,891.0000
Survived,1.0000
Pclass,3.0000
Age,80.0000
SibSp,8.0000
Parch,6.0000
Fare,512.3292


###**Method 7 :** Pandas min() Method

In [ ]:
titanic.min(numeric_only = True)

,0
PassengerId,1.00
Survived,0.00
Pclass,1.00
Age,0.42
SibSp,0.00
Parch,0.00
Fare,0.00


###**Method 8 :** Pandas size Method.
###*(Returns size of dataframe/series which is equivalent to total number of elements. That is rows x columns).*

In [ ]:
titanic.size

10692

###**Method 9 :** Pandas shape Method
###*(Returns tuple of shape (Rows, columns) of dataframe/series)*

In [ ]:
titanic.shape

(891, 12)

###**Method 10 :** Pandas Ndim Method
###*Returns dimension of dataframe/series. 1 for one dimension (series), 2 for two dimension (dataframe)*

In [ ]:
titanic.ndim

2

###**Method 11 :** pandas dtype method
###*this method returns the datatype of all the columns of a dataframe*

In [ ]:
titanic.dtypes

,0
PassengerId,int64
Survived,int64
Pclass,int64
Name,object
Sex,object
Age,float64
SibSp,int64
Parch,int64
Ticket,object
Fare,float64


##Print out the number who survived

In [ ]:
No_Of_People_Survived=titanic['Survived'].sum()
print("The Number of People that Survived the Shipwreck is %d"%(No_Of_People_Survived))

The Number of People that Survived the Shipwreck is 342


#Transforming the Dataset

###Dropping The Irrelevant Column

In [ ]:
titanic.isnull().sum()

,0
PassengerId,0
Survived,0
Pclass,0
Name,0
Sex,0
Age,177
SibSp,0
Parch,0
Ticket,0
Fare,0


From our prelimnary analysis we can see that the embarked column has two missing value, as it's too low, we can just simply drop those rows.

In [ ]:
titanic.shape

(891, 12)

In [ ]:
titanic=titanic.dropna(subset=['Embarked'])


In [ ]:
titanic.shape

(889, 12)

We can see that the Column "Cabin" has lots of missing values, also as the cabin column doesnt add relevance to the data we can just go ahead and drop the entire column.

In [ ]:
 titanic.drop('Cabin',inplace=True, axis=1)

<ipython-input-103-c7bd0849db29>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  titanic.drop('Cabin',inplace=True, axis=1)


In [ ]:
titanic

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,S
...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C


Now the Only Column that we are left with to deal with missing values is "Age". It's an important feature for analysis, so we cant drop it and we cant just drop the missing values as they make up of 1/8Th of the Data. So we Impute the data
We will create two seprate datasets

1} where Age column is Imputed by Mean

2} where Missing Values are dropped.

In [ ]:
titanic_age=titanic.dropna()

In [ ]:
titanic_age

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,S
...,...,...,...,...,...,...,...,...,...,...,...
885,886,0,3,"Rice, Mrs. William (Margaret Norton)",female,39.0,0,5,382652,29.1250,Q
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C


In [ ]:
from sklearn.impute import SimpleImputer
imr = SimpleImputer(missing_values='NaN', strategy='mean')
imr = imr.fit(titanic)
titanic_imputed = imr.transform(titanic.values)

ValueError: Cannot use mean strategy with non-numeric data:
could not convert string to float: 'Braund, Mr. Owen Harris'

We can see that the we cant impute the data until we convert the non-nummeric data into numerical encoding. So our next step is to Drop the irrelevant columns and Convert Non-Numeric columns into Numerical columns.

###**Columns to Drop**
###**PassengerId** : *We dont need passenger id as it contains distinct vales which are not useful in analysis.*
### **Passenger name** : *We don't need passenger name as we are not focussing on one particular person.*
###**Ticket** : *We don't Need Ticket Number as it Contains Distinct values and won't be Helpful for Training any Model.*
###**Fare** : *It contains Distinct Values and the range of the Fare cannot be Defined, Also Pclass Column gives us a Straightforward idea about Passengers Ticket Price, As for 3Rd Class price is around 7. While for 1St class is more than 30.*

In [ ]:
 titanic_imputed =titanic.drop(['PassengerId','Name','Ticket','Fare'], axis=1)
 titanic_age=titanic_age.drop(['PassengerId','Name','Ticket','Fare'], axis=1)

In [ ]:
titanic_imputed

,Survived,Pclass,Sex,Age,SibSp,Parch,Embarked
0,0,3,male,22.0,1,0,S
1,1,1,female,38.0,1,0,C
2,1,3,female,26.0,0,0,S
3,1,1,female,35.0,1,0,S
4,0,3,male,35.0,0,0,S
...,...,...,...,...,...,...,...
886,0,2,male,27.0,0,0,S
887,1,1,female,19.0,0,0,S
888,0,3,female,NaN,1,2,S
889,1,1,male,26.0,0,0,C


###Encoding categorical data using one-hot encoding.

In [ ]:
titanic_imputed = pd.get_dummies(titanic_imputed, columns = ['Sex', 'Embarked'])
titanic_imputed

,Survived,Pclass,Age,SibSp,Parch,Sex_female,Sex_male,Embarked_C,Embarked_Q,Embarked_S
0,0,3,22.0,1,0,False,True,False,False,True
1,1,1,38.0,1,0,True,False,True,False,False
2,1,3,26.0,0,0,True,False,False,False,True
3,1,1,35.0,1,0,True,False,False,False,True
4,0,3,35.0,0,0,False,True,False,False,True
...,...,...,...,...,...,...,...,...,...,...
886,0,2,27.0,0,0,False,True,False,False,True
887,1,1,19.0,0,0,True,False,False,False,True
888,0,3,NaN,1,2,True,False,False,False,True
889,1,1,26.0,0,0,False,True,True,False,False


Now as we have got our relevant columns and all the Data is Numerical, we can now Impute the Missing Age Values

In [ ]:
from sklearn.impute import SimpleImputer
import numpy as np
imr = SimpleImputer(missing_values=np.nan, strategy='median')
titanic_imputed.isna().sum()

,0
Survived,0
Pclass,0
Age,177
SibSp,0
Parch,0
Sex_female,0
Sex_male,0
Embarked_C,0
Embarked_Q,0
Embarked_S,0


In [ ]:
imr = imr.fit(titanic_imputed)
titanic_imputed = imr.transform(titanic_imputed.values)

/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but SimpleImputer was fitted with feature names
  warnings.warn(


In [ ]:
titanic_imputed_df = pd.DataFrame(titanic_imputed,columns=["Survived", "Pclass",	 "Age",	"SibSp",	"Parch",	"Sex_female",	"Sex_male","Embarked_C","Embarked_Q","Embarked_S"])

In [ ]:
titanic_imputed_df.isna().sum()

,0
Survived,0
Pclass,0
Age,0
SibSp,0
Parch,0
Sex_female,0
Sex_male,0
Embarked_C,0
Embarked_Q,0
Embarked_S,0


#Exploratory data Analysis and Hypothesis Testing

####Dataset 1 : Original data with Columns dropped and Age vales with "NaN' dropped

In [ ]:
titanic_age

,Survived,Pclass,Sex,Age,SibSp,Parch,Embarked
0,0,3,male,22.0,1,0,S
1,1,1,female,38.0,1,0,C
2,1,3,female,26.0,0,0,S
3,1,1,female,35.0,1,0,S
4,0,3,male,35.0,0,0,S
...,...,...,...,...,...,...,...
885,0,3,female,39.0,0,5,Q
886,0,2,male,27.0,0,0,S
887,1,1,female,19.0,0,0,S
889,1,1,male,26.0,0,0,C


####Dataset 2 : Transformed Dataset

In [ ]:
titanic_imputed_df

,Survived,Pclass,Age,SibSp,Parch,Sex_female,Sex_male,Embarked_C,Embarked_Q,Embarked_S
0,0.0,3.0,22.0,1.0,0.0,0.0,1.0,0.0,0.0,1.0
1,1.0,1.0,38.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0
2,1.0,3.0,26.0,0.0,0.0,1.0,0.0,0.0,0.0,1.0
3,1.0,1.0,35.0,1.0,0.0,1.0,0.0,0.0,0.0,1.0
4,0.0,3.0,35.0,0.0,0.0,0.0,1.0,0.0,0.0,1.0
...,...,...,...,...,...,...,...,...,...,...
884,0.0,2.0,27.0,0.0,0.0,0.0,1.0,0.0,0.0,1.0
885,1.0,1.0,19.0,0.0,0.0,1.0,0.0,0.0,0.0,1.0
886,0.0,3.0,28.0,1.0,2.0,1.0,0.0,0.0,0.0,1.0
887,1.0,1.0,26.0,0.0,0.0,0.0,1.0,1.0,0.0,0.0


###**Hypothesis :**
*In the Movie it's shown that the Women & children were given first preferrence in evacuations while the ship sank, so from that point, there has to be more Female Survivors along with childrens. So lets test it*.

**Let's Start with dataset 1**

In [ ]:
Test1=titanic_age.groupby(['Survived'])
Test1.first()

,Pclass,Sex,Age,SibSp,Parch,Embarked
Survived,,,,,,
0,3,male,22.0,1,0,S
1,1,female,38.0,1,0,C


In [ ]:
Test2=Test1.get_group(1)
Female=Test2['Sex'].value_counts()['female']
Male=Test2['Sex'].value_counts()['male']
print("Number of Female Passengers That Survived : %d"%Female)
print("Number of Male Passengers That Survived : %d"%Male)

Number of Female Passengers That Survived : 195
Number of Male Passengers That Survived : 93


<ipython-input-118-15328476598a>:1: FutureWarning: When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.
  Test2=Test1.get_group(1)


We can clearly see that our Hypothesis Holds True. There are more Female Survivors than Male.**

Now lets Check, of all male that survived how many where children

In [ ]:
Test3=Test2[Test2['Sex']=='male']
Test4=Test3[Test3['Age']<=18.0]
Test4.count()

,0
Survived,24
Pclass,24
Sex,24
Age,24
SibSp,24
Parch,24
Embarked,24


In [ ]:
Test1.count()

,Pclass,Sex,Age,SibSp,Parch,Embarked
Survived,,,,,,
0,424,424,424,424,424,424
1,288,288,288,288,288,288


####Conclusion (Dataset 1):
 *Altogther we have 889 records of people of Which we dropped 177 record with missing Age, so we now have 288 Survived and 424 Dead passenger samples. Out of 288 Passengers that survived 195 were female and 93 were male. Out of all the Male passengers that survived 24 were children and 69 Adults. So we can see that our Hypothesis is correct. As the Women and childrens were evacuated early Their Population is much more as compared to male population that survided.*

**Let's Check with Second Dataset**

In [ ]:
Test_Imp_1=titanic_imputed_df.groupby(['Survived'])
Test_Imp_1.first()

,Pclass,Age,SibSp,Parch,Sex_female,Sex_male,Embarked_C,Embarked_Q,Embarked_S
Survived,,,,,,,,,
0.0,3.0,22.0,1.0,0.0,0.0,1.0,0.0,0.0,1.0
1.0,1.0,38.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0


In [ ]:
Test_Imp_2=Test_Imp_1.get_group(1)
Female=int(Test_Imp_2['Sex_female'].sum())
Male=int(Test_Imp_2['Sex_male'].sum())
print("Number of Female Passengers That Survived : %d"%Female)
print("Number of Male Passengers That Survived : %d"%Male)

Number of Female Passengers That Survived : 231
Number of Male Passengers That Survived : 109


<ipython-input-122-b1f8b7f19af7>:1: FutureWarning: When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.
  Test_Imp_2=Test_Imp_1.get_group(1)


*We can clearly see that our Hypothesis Holds True. There are more Female Survivors than Male.*

Now lets Check, of all male that survived how many where children

In [ ]:
Test_Imp_3=Test_Imp_2[Test_Imp_2['Sex_male']==1.0]
Test_Imp_4=Test_Imp_3[Test_Imp_3['Age']<=18.0]
Test_Imp_4.count()

,0
Survived,24
Pclass,24
Age,24
SibSp,24
Parch,24
Sex_female,24
Sex_male,24
Embarked_C,24
Embarked_Q,24
Embarked_S,24


In [ ]:
Test_Imp_1.count()

,Pclass,Age,SibSp,Parch,Sex_female,Sex_male,Embarked_C,Embarked_Q,Embarked_S
Survived,,,,,,,,,
0.0,549,549,549,549,549,549,549,549,549
1.0,340,340,340,340,340,340,340,340,340


####**Conclusion (Dataset 2):**
Altogther we have 889 records of people so we now have 340 Survived and 549 Dead passenger samples. Out of 340 Passengers that survived 231 were female and 109 were male. Out of all the Male passengers that survived 24 were children and 85 Adults. So we can see that our Hypothesis is correct. As the Women and childrens were evacuated early Their Population is much more as compared to male population that survided.

####Explore creating a new quantized feature for numeric data

We can Try another method where we can create bins for age groups

In [ ]:
bins = [0, 10, 18, 65, np.inf]
names = ['Child', 'Teen', 'Adult', 'Old']

titanic_age['AgeRange'] = pd.cut(titanic_age['Age'], bins, labels=names)

print(titanic_age.dtypes)

Survived       int64
Pclass         int64
Sex           object
Age          float64
SibSp          int64
Parch          int64
Embarked      object
AgeRange    category
dtype: object


In [ ]:
titanic_age

,Survived,Pclass,Sex,Age,SibSp,Parch,Embarked,AgeRange
0,0,3,male,22.0,1,0,S,Adult
1,1,1,female,38.0,1,0,C,Adult
2,1,3,female,26.0,0,0,S,Adult
3,1,1,female,35.0,1,0,S,Adult
4,0,3,male,35.0,0,0,S,Adult
...,...,...,...,...,...,...,...,...
885,0,3,female,39.0,0,5,Q,Adult
886,0,2,male,27.0,0,0,S,Adult
887,1,1,female,19.0,0,0,S,Adult
889,1,1,male,26.0,0,0,C,Adult


####**Either we can get a categorical value or we can Convert This ordinal column into a numeric datatype**

In [ ]:
titanic_age=titanic_age.dropna(subset=['AgeRange'])

In [ ]:
names = [0, 1, 2, 3]

titanic_age['AgeRange'] = pd.cut(titanic_age['Age'], bins, labels=names)
titanic_age["AgeRange"] = titanic_age["AgeRange"].astype("int64")
print(titanic_age.dtypes)

Survived      int64
Pclass        int64
Sex          object
Age         float64
SibSp         int64
Parch         int64
Embarked     object
AgeRange      int64
dtype: object


In [ ]:
titanic_age

,Survived,Pclass,Sex,Age,SibSp,Parch,Embarked,AgeRange
0,0,3,male,22.0,1,0,S,2
1,1,1,female,38.0,1,0,C,2
2,1,3,female,26.0,0,0,S,2
3,1,1,female,35.0,1,0,S,2
4,0,3,male,35.0,0,0,S,2
...,...,...,...,...,...,...,...,...
885,0,3,female,39.0,0,5,Q,2
886,0,2,male,27.0,0,0,S,2
887,1,1,female,19.0,0,0,S,2
889,1,1,male,26.0,0,0,C,2


**This method Works Well for data Anlysis but would not give us better result when we go for Classification, regression, or clustering Methods. So we can try One Hot Encoding Instead**

In [ ]:
bins = [0, 10, 18, 65, np.inf]
names = ['Child', 'Teen', 'Adult', 'Old']

titanic_imputed_df['AgeRange'] = pd.cut(titanic_imputed_df['Age'], bins, labels=names)

In [ ]:
titanic_imputed_df = pd.get_dummies(titanic_imputed_df, columns = ['AgeRange'])
titanic_imputed_df

,Survived,Pclass,Age,SibSp,Parch,Sex_female,Sex_male,Embarked_C,Embarked_Q,Embarked_S,AgeRange_Child,AgeRange_Teen,AgeRange_Adult,AgeRange_Old
0,0.0,3.0,22.0,1.0,0.0,0.0,1.0,0.0,0.0,1.0,False,False,True,False
1,1.0,1.0,38.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,False,False,True,False
2,1.0,3.0,26.0,0.0,0.0,1.0,0.0,0.0,0.0,1.0,False,False,True,False
3,1.0,1.0,35.0,1.0,0.0,1.0,0.0,0.0,0.0,1.0,False,False,True,False
4,0.0,3.0,35.0,0.0,0.0,0.0,1.0,0.0,0.0,1.0,False,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
884,0.0,2.0,27.0,0.0,0.0,0.0,1.0,0.0,0.0,1.0,False,False,True,False
885,1.0,1.0,19.0,0.0,0.0,1.0,0.0,0.0,0.0,1.0,False,False,True,False
886,0.0,3.0,28.0,1.0,2.0,1.0,0.0,0.0,0.0,1.0,False,False,True,False
887,1.0,1.0,26.0,0.0,0.0,0.0,1.0,1.0,0.0,0.0,False,False,True,False


In [ ]:
childrens=len(titanic_imputed_df.loc[(titanic_imputed_df['AgeRange_Child']==True) & (titanic_imputed_df['Survived']==1.0) & (titanic_imputed_df['Sex_male']==1.0)])
women=len(titanic_imputed_df.loc[(titanic_imputed_df['Sex_female']==1.0) & (titanic_imputed_df['Survived']==1.0)])
total_women_child=childrens+women
total_women_child

250

#References :-
https://www.geeksforgeeks.org/get-first-n-records-of-a-pandas-dataframe/

https://www.geeksforgeeks.org/python-extracting-rows-using-pandas-iloc/

https://statisticsglobe.com/summary-statistics-pandas-dataframe-python

https://www.geeksforgeeks.org/python-pandas-df-size-df-shape-and-df-ndim/

https://www.askpython.com/python-modules/pandas/remove-column-from-python-dataframe#:~:text=How%20to%20Remove%20a%20Column%20From%20a%20Python,Python%20drop%20%28%29%20function%20to%20remove%20a%20column

https://www.datasciencelearner.com/convert-numpy-array-to-dataframe/#:~:text=Steps%20by%20Steps%20to%20convert%20Numpy%20array%20to,3%3A%20Convert%20the%20numpy%20array%20to%20the%20dataframe.

https://stackoverflow.com/questions/64412842/how-to-fix-simpleimputer-valueerror-input-contains-nan-infinity-or-a-value-t

https://stackoverflow.com/questions/49382207/how-to-map-numeric-data-into-categories-bins-in-pandas-dataframe